addons/cookoff/functions/fnc_cookOffEffect.sqf

```diff
@@ -20,60 +20,89 @@
 
 private _sound = objNull;
+private _fireKey = "";
+
 if (isServer) then {
-    // ironically biggest performance hit is this. Creating a new sound source takes up aprox 400 milliseconds.
-    // I dont think there is an alternative that takes into effect distance and whatever, but if you find one please fix!
+    // Spawn sound effect
     if (_jet || _ring) then {
         private _soundName = selectRandomWeighted [QGVAR(Sound_low), 0.1, QGVAR(Sound_mid), 0.25, QGVAR(Sound_high), 0.65];
-        _sound = createSoundSource [_soundName, position _obj, [], 0];
+        _sound = createSoundSource [_soundName, ASLToAGL getPosASL _vehicle, [], 0];
+        _sound attachTo [_vehicle];
     };
 
-    if (_ring) then {
-        private _intensity = 6;
-        private _radius = 1.5 * ((boundingBoxReal _obj) select 2);
-        [QEGVAR(fire,addFireSource), [_obj, _radius, _intensity, format [QGVAR(%1), hashValue _obj]]] call CBA_fnc_localEvent;
+    // Make the ring a source of fire
+    if (_ring && {["ace_fire"] call EFUNC(common,isModLoaded)}) then {
+        _fireKey = format [QGVAR(%1), hashValue _vehicle];
+
+        [QEGVAR(fire,addFireSource), [_vehicle, FLAME_SIZE * ((boundingBoxReal _vehicle) select 2), FIRE_INTENSITY, _fireKey]] call CBA_fnc_localEvent;
     };
 };
 
 [{
-    params ["_args", "_pfh"];
-    _args params ["_obj", "_jet", "_ring", "_time", "_startTime", "_light", "_fireSelection", "_sound", "_intensity"];
+    (_this select 0) params ["_vehicle", "_jet", "_ring", "_duration", "_startTime", "_light", "_fireSelection", "_sound", "_intensity", "_fireKey"];
+
     private _elapsedTime = CBA_missionTime - _startTime;
-    if (_elapsedTime >= _time) exitWith {
+
+    // Clean up effects once effects have finished or vehicle has been deleted
+    if (isNull _vehicle || {_elapsedTime >= _duration}) exitWith {
+        [_this select 1] call CBA_fnc_removePerFrameHandler;
+
         deleteVehicle _light;
-        deleteVehicle _sound;
+
         if (isServer) then {
-            [QEGVAR(fire,removeFireSource), [format [QGVAR(%1), hashValue _obj]]] call CBA_fnc_localEvent;
+            deleteVehicle _sound;
+
+            if (["ace_fire"] call EFUNC(common,isModLoaded)) then {
+                [QEGVAR(fire,removeFireSource), _fireKey] call CBA_fnc_localEvent;
+            };
         };
-        [_pfh] call CBA_fnc_removePerFrameHandler;
     };
-    private _factor = (1 + (_elapsedTime / 2) min 2);
-    private _flameSize = 1.5;
 
-    if (_elapsedTime > (_time * (3 / 4))) then {
-        _factor = _factor * linearConversion [_time * (3 / 4), _time, _elapsedTime, 1, 0.5];
+    private _factor = 1 + (_elapsedTime / 2) min 2;
+
+    if (_elapsedTime > _duration * 3 / 4) then {
+        _factor = _factor * linearConversion [_duration * 3 / 4, _duration, _elapsedTime, 1, 0.5];
+    };
+
+    // Make flame push object into ground to make effect seem more "alive"
+    if (_jet && !isGamePaused && {local _vehicle}) then {
```

Comment:
This was a problem before your refactor, but the force addition here is tied to framerate instead of game time. This results in undesirable behavior when time acceleration isn't equal to one. It's a particular problem for factors less than 0.1, including a factor of 0, which doesn't cause `isGamePaused` to return true. 
There is probably a better approach, but a solution I've been doing locally is this:
```sqf 
if (_jet && !isGamePaused && { local _obj }) then {
    private _nextForceTime = _obj getVariable [QGVAR(nextForceTime), 0];
    if (_nextFlameTime > CBA_missionTime) then {
        private _force = [0, 0, _factor * -(0.5 min random 1.5) * (0.3 min random 1)] vectorMultiply getMass _obj;
        _obj addForce [_force, vectorUpVisual _obj];
        _obj setVariable [QGVAR(nextForceTime), _nextForceTime + 0.015]; // 15ms/circa 67 Hz limit
    };
};
```

Reply:
> undesirable behavior
Can you elaborate on that?

Reply:
Sorry for not being clearer. When time acceleration is small or 0 (such as when the splendid camera is first opened, or in some slowed events), vehicles cooking off (if `_jet` is true) still recieve a force down every frame. The result of this per-frame force can be a number of things but often vehicles will either have a sudden collision with the ground and send themselves flying or clip through all together, i.e.,

https://github.com/acemod/ACE3/assets/11843718/e11d02c8-7104-47ba-9186-0c1a1ac794e2
Note time acceleration in top right.

Reply:
Thanks!